### ISSUE TITLE
NoClassDefFoundError: javax/annotation/Generated on Java 9

### ISSUE THREAD

**Issue by davido** [google/auto#560]
//CC @cushon 

This is the very first time, I'm trying to build [Gerrit Code Review](https://gerrit.googlesource.com/gerrit) with Bazel, with Java 9. We have this Bazel patch: https://github.com/hhclam/bazel/commit/64212c8026df9fc6361d5af8414acc373e221955.

One issue is EP doesn't really work with mentioned Bazel patch, the vanilla javac is used and not the bundled one with Bazel:

```
ERROR: /home/davido/projects/gerrit2/java/com/google/gerrit/common/BUILD:32:1: Building java/com/google/gerrit/common/libserver.jar (57 source files) failed (Exit 1)
java/com/google/gerrit/common/RawInputUtil.java:34: error: [CheckReturnValue] Ignored return value of method that is annotated with @CheckReturnValue
    Preconditions.checkNotNull(bytes);
                              ^
    (see http://errorprone.info/bugpattern/CheckReturnValue)
  Did you mean to remove this line?
[...]
ERROR: /home/davido/projects/gerrit2/java/com/google/gerrit/util/cli/BUILD:1:1: Building java/com/google/gerrit/util/cli/libcli.jar (6 source files) failed (Exit 1)
java/com/google/gerrit/util/cli/CmdLineParser.java:229: error: [CheckReturnValue] Ignored return value of method that is annotated with @CheckReturnValue
        map.put(ent.getKey(), val);
               ^
    (see http://errorprone.info/bugpattern/CheckReturnValue)
  Did you mean to remove this line?
Target //java/com/google/gerrit/server:server failed to build
```

EP can be disabled for now, using: --javacopt="-XepDisableAllChecks" option.

Gerrit issue, is outdated auto-value version used (1.4.1). That's because the `auto-value` upgrade had to be reverted due to Eclipse-Oxygen bug, that was fixed meantime, but wasn't released yet. 

Auto-value is upgraded with this diff (applied on gerrit master):

```
diff --git a/WORKSPACE b/WORKSPACE
index 6810f79c93..4f5de9b620 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -408,8 +408,8 @@ maven_jar(
 
 maven_jar(
     name = "auto_value",
-    artifact = "com.google.auto.value:auto-value:1.4.1",
-    sha1 = "8172ebbd7970188aff304c8a420b9f17168f6f48",
+    artifact = "com.google.auto.value:auto-value:1.5.2",
+    sha1 = "1b94ab7ec707e2220a0d1a7517488d1843236345",
 )
 
 maven_jar(
```

With these three fixes:

- Custom Bazel build
- Disabled EP
- Auto-Value upgrade to 1.5.2

trying to build `java/com/google/gerrit/sshd:sshd` on Java 9 is failing with:

```
$ /home/davido/projects/bazel/bazel-bin/src/bazel build java/com/google/gerrit/sshd:sshd --javacopt="-XepDisableAllChecks" --host_java_toolchain=@bazel_tools//tools/jdk:jdk9 --java_toolchain=@bazel_tools//tools/jdk:jdk9
INFO: Analysed target //java/com/google/gerrit/sshd:sshd (0 packages loaded).
INFO: Found 1 target...
ERROR: /home/davido/projects/gerrit/java/com/google/gerrit/sshd/BUILD:1:1: Building java/com/google/gerrit/sshd/libsshd.jar (97 source files) and running annotation processors (AutoAnnotationProcessor, AutoValueProcessor) failed (Exit 1)
java.lang.RuntimeException: java.lang.NoClassDefFoundError: javax/annotation/Generated
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:158)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
	at com.google.devtools.build.buildjar.javac.BlazeJavacMain.compile(BlazeJavacMain.java:107)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder$1.invokeJavac(SimpleJavaLibraryBuilder.java:105)
	at com.google.devtools.build.buildjar.ReducedClasspathJavaLibraryBuilder.compileSources(ReducedClasspathJavaLibraryBuilder.java:54)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.compileJavaLibrary(SimpleJavaLibraryBuilder.java:108)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.run(SimpleJavaLibraryBuilder.java:116)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.processRequest(BazelJavaBuilder.java:105)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.runPersistentWorker(BazelJavaBuilder.java:67)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.main(BazelJavaBuilder.java:45)
Caused by: java.lang.NoClassDefFoundError: javax/annotation/Generated
	at com.google.auto.value.processor.AutoAnnotationProcessor.getReferencedTypes(AutoAnnotationProcessor.java:450)
	at com.google.auto.value.processor.AutoAnnotationProcessor.processMethod(AutoAnnotationProcessor.java:152)
	at com.google.auto.value.processor.AutoAnnotationProcessor.process(AutoAnnotationProcessor.java:129)
	at com.google.auto.value.processor.AutoAnnotationProcessor.process(AutoAnnotationProcessor.java:113)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:968)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:884)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.access$2200(JavacProcessingEnvironment.java:108)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1206)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1315)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1246)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:922)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
	... 10 more
Caused by: java.lang.ClassNotFoundException: javax.annotation.Generated
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:466)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:563)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:496)
	... 23 more
Target //java/com/google/gerrit/sshd:sshd failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.626s, Critical Path: 0.43s
FAILED: Build did NOT complete successfully
```

Environment: Linux.

```
$ java -version
openjdk version "9.0.1"
```

**Comment by davido** [comment]
I went ahead and added 'javax.annotation:javax.annotation-api:jar:1.3.1' to Gerrit with this patch: [1]. Now I'm, getting a different error:

```
$ /home/davido/projects/bazel/bazel-bin/src/bazel build java/com/google/gerrit/sshd:sshd  --javacopt="-XepDisableAllChecks" --host_java_toolchain=@bazel_tools//tools/jdk:jdk9 --java_toolchain=@bazel_tools//tools/jdk:jdk9
INFO: Analysed target //java/com/google/gerrit/sshd:sshd (1 packages loaded).
INFO: Found 1 target...
ERROR: /home/davido/projects/gerrit/java/com/google/gerrit/sshd/BUILD:1:1: Building java/com/google/gerrit/sshd/libsshd.jar (97 source files) and running annotation processors (AutoAnnotationProcessor, AutoValueProcessor) failed (Exit 1)
java/com/google/gerrit/sshd/Commands.java:48: error: @AutoAnnotation processor threw an exception: java.lang.NullPointerException
  public static CommandName named(String value) {
                            ^
  	at com.google.auto.value.processor.AutoAnnotationProcessor.getTypeMirror(AutoAnnotationProcessor.java:470)
  	at com.google.auto.value.processor.AutoAnnotationProcessor.getReferencedTypes(AutoAnnotationProcessor.java:450)
  	at com.google.auto.value.processor.AutoAnnotationProcessor.processMethod(AutoAnnotationProcessor.java:152)
  	at com.google.auto.value.processor.AutoAnnotationProcessor.process(AutoAnnotationProcessor.java:129)
  	at com.google.auto.value.processor.AutoAnnotationProcessor.process(AutoAnnotationProcessor.java:113)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:968)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:884)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.access$2200(JavacProcessingEnvironment.java:108)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1206)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1315)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1246)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:922)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
  	at com.google.devtools.build.buildjar.javac.BlazeJavacMain.compile(BlazeJavacMain.java:107)
  	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder$1.invokeJavac(SimpleJavaLibraryBuilder.java:105)
  	at com.google.devtools.build.buildjar.ReducedClasspathJavaLibraryBuilder.compileSources(ReducedClasspathJavaLibraryBuilder.java:54)
  	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.compileJavaLibrary(SimpleJavaLibraryBuilder.java:108)
  	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.run(SimpleJavaLibraryBuilder.java:116)
  	at com.google.devtools.build.buildjar.BazelJavaBuilder.processRequest(BazelJavaBuilder.java:105)
  	at com.google.devtools.build.buildjar.BazelJavaBuilder.runPersistentWorker(BazelJavaBuilder.java:67)
  	at com.google.devtools.build.buildjar.BazelJavaBuilder.main(BazelJavaBuilder.java:45)
java.lang.RuntimeException: java.lang.NullPointerException
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:158)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
	at com.google.devtools.build.buildjar.javac.BlazeJavacMain.compile(BlazeJavacMain.java:107)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder$1.invokeJavac(SimpleJavaLibraryBuilder.java:105)
	at com.google.devtools.build.buildjar.ReducedClasspathJavaLibraryBuilder.compileSources(ReducedClasspathJavaLibraryBuilder.java:54)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.compileJavaLibrary(SimpleJavaLibraryBuilder.java:108)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.run(SimpleJavaLibraryBuilder.java:116)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.processRequest(BazelJavaBuilder.java:105)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.runPersistentWorker(BazelJavaBuilder.java:67)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.main(BazelJavaBuilder.java:45)
Caused by: java.lang.NullPointerException
	at com.google.auto.value.processor.AutoAnnotationProcessor.getTypeMirror(AutoAnnotationProcessor.java:470)
	at com.google.auto.value.processor.AutoAnnotationProcessor.getReferencedTypes(AutoAnnotationProcessor.java:450)
	at com.google.auto.value.processor.AutoAnnotationProcessor.processMethod(AutoAnnotationProcessor.java:152)
	at com.google.auto.value.processor.AutoAnnotationProcessor.process(AutoAnnotationProcessor.java:129)
	at com.google.auto.value.processor.AutoAnnotationProcessor.process(AutoAnnotationProcessor.java:113)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:968)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:884)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.access$2200(JavacProcessingEnvironment.java:108)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1206)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1315)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1246)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:922)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
	... 10 more
Target //java/com/google/gerrit/sshd:sshd failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 18.535s, Critical Path: 18.33s
FAILED: Build did NOT complete successfully
```

[1] http://paste.openstack.org/show/628224

**Comment by davido** [comment]
The NPE seems to be patched Bazel related. I [extracted this reproducer](https://github.com/davido/ostrovsky) from gerrit Code Review project and verified that auto-value jst worked with Maven, but failing with NPE with Bazel.

**Comment by cushon** [comment]
The part of this that is an auto bug is that `AutoAnnotationProcessor` should be using something like c9fd48f5f8e80651d30173f2113089bce68955ff instead of trying to load `javax.annotatation.Generated` here:  https://github.com/google/auto/blob/c9fd48f5f8e80651d30173f2113089bce68955ff/value/src/main/java/com/google/auto/value/processor/AutoAnnotationProcessor.java#L450

That avoids the need for an explicit dep on `javax.annotation.Generated`, and fixes:

```
Caused by: java.lang.NoClassDefFoundError: javax/annotation/Generated
	at com.google.auto.value.processor.AutoAnnotationProcessor.getReferencedTypes(AutoAnnotationProcessor.java:450)
```

The other NPE:

```
Caused by: java.lang.NullPointerException
	at com.google.auto.value.processor.AutoAnnotationProcessor.getTypeMirror(AutoAnnotationProcessor.java:470)
	at com.google.auto.value.processor.AutoAnnotationProcessor.getReferencedTypes(AutoAnnotationProcessor.java:450)
```

... appears to be because the patch to gerrit and the repro at davido/ostrovsky add `javax-annotation` to the processor classpath, but not the compilation classpath. (I'm not sure maven makes that distinction, which might be why it works?) Anyway, exporting `javax-annotation` from the `auto-value` target avoids the NPE:

```
diff --git a/lib/BUILD b/lib/BUILD
index 242d328..39eaabf 100644
--- a/lib/BUILD
+++ b/lib/BUILD
@@ -28,5 +28,8 @@ java_library(
         ":auto-value-plugin",
     ],
     visibility = ["//visibility:public"],
-    exports = ["@auto_value//jar"],
+    exports = [
+        ":javax-annotation",
+        "@auto_value//jar",
+    ],
 )
```

**Comment by tbroyer** [comment]
> I'm not sure maven makes that distinction, which might be why it works?

Maven **can** make the distinction (https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#annotationProcessorPaths) but the repro case here does not use the processor classpath (i.e. annotation processors are looked up in the compilation classpath).
So I would say this is why it works.

**Comment by davido** [comment]
@cushon Thanks. It was too early today morning. I can build GWT UI based on Java 9 now. JS stack is still failing with Dagger related stuff, used by rules_closure, reported here [1].

Even more interesting, if I try to install gerrit.war, that was built with Java 9, I'm running into NPE:

```
Exception in thread "main" java.lang.ExceptionInInitializerError
	at com.google.gerrit.server.index.change.ChangeSchemaDefinitions.<clinit>(ChangeSchemaDefinitions.java:25)
	at com.google.gerrit.server.index.IndexModule.<clinit>(IndexModule.java:78)
	at com.google.gerrit.pgm.init.InitIndex.run(InitIndex.java:69)
	at com.google.gerrit.pgm.init.SitePathInitializer.run(SitePathInitializer.java:93)
	at com.google.gerrit.pgm.init.BaseInit.run(BaseInit.java:138)
	at com.google.gerrit.pgm.util.AbstractProgram.main(AbstractProgram.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at com.google.gerrit.launcher.GerritLauncher.invokeProgram(GerritLauncher.java:204)
	at com.google.gerrit.launcher.GerritLauncher.mainImpl(GerritLauncher.java:109)
	at com.google.gerrit.launcher.GerritLauncher.main(GerritLauncher.java:64)
	at Main.main(Main.java:24)
Caused by: java.lang.NullPointerException: Null rule
	at com.google.gerrit.server.project.AutoValue_SubmitRuleOptions$Builder.rule(AutoValue_SubmitRuleOptions.java:102)
	at com.google.gerrit.server.project.SubmitRuleOptions.builder(SubmitRuleOptions.java:32)
	at com.google.gerrit.server.index.change.ChangeField.<clinit>(ChangeField.java:566)
	... 14 more
```

* [1] https://github.com/bazelbuild/rules_closure/issues/234

**Comment by davido** [comment]
@cushon Thanks for the pointer.

> The part of this that is an auto bug is that AutoAnnotationProcessor should be using something like c9fd48f instead of trying to load javax.annotatation.Generated here:

I was able to replace this hard coded part with:

```
diff --git a/value/src/main/java/com/google/auto/value/processor/AutoAnnotationProcessor.java b/value/src/main/java/com/google/auto/value/processor/AutoAnnotationProcessor.java
index 4a7780d..0d68f95 100644
--- a/value/src/main/java/com/google/auto/value/processor/AutoAnnotationProcessor.java
+++ b/value/src/main/java/com/google/auto/value/processor/AutoAnnotationProcessor.java
@@ -17,6 +17,7 @@ package com.google.auto.value.processor;
 
 import static java.util.stream.Collectors.toCollection;
 
+import com.google.auto.common.GeneratedAnnotations;
 import com.google.auto.common.MoreElements;
 import com.google.auto.common.SuperficialValidation;
 import com.google.auto.service.AutoService;
@@ -35,7 +36,6 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
-import javax.annotation.Generated;
 import javax.annotation.processing.AbstractProcessor;
 import javax.annotation.processing.ProcessingEnvironment;
 import javax.annotation.processing.Processor;
@@ -191,12 +191,12 @@ public class AutoAnnotationProcessor extends AbstractProcessor {
   }
 
   private String getGeneratedTypeName(TypeSimplifier typeSimplifier) {
-    TypeElement generatedTypeElement =
-        processingEnv.getElementUtils().getTypeElement("javax.annotation.Generated");
-    if (generatedTypeElement == null) {
+    Optional<TypeElement> generatedTypeElement =
+        GeneratedAnnotations.generatedAnnotation(processingEnv.getElementUtils());
+    if (!generatedTypeElement.isPresent()) {
       return "";
     }
-    return typeSimplifier.simplify(generatedTypeElement.asType());
+    return typeSimplifier.simplify(generatedTypeElement.get().asType());
   }
 
   /**
```

But, in the same compilation unit we have this:

```
  private Set<TypeMirror> getReferencedTypes(
      TypeMirror annotation,
      ExecutableElement method,
      Set<TypeMirror> memberTypes,
      Set<Class<?>> wrapperTypesUsedInCollections) {
    Set<TypeMirror> types = new TypeMirrorSet();
    types.add(annotation);
    types.add(getTypeMirror(Generated.class));
[...]
```

that still references `javax.annotation.Generated´ that is missing in Java 9. Any clue, what is the approach to replace that part on Java 9? Thanks!

**Comment by davido** [comment]
Seems to be fixed with: f04406c1f11d3796bd93ed89ae160f47bcb4b7e4.
